source/server/http/admin.cc

```diff
@@ -520,17 +551,86 @@ Http::Code AdminImpl::runCallback(const std::string& path, Buffer::Instance& res
     for (auto handler : sorted_handlers) {
       response.add(fmt::format("  {}: {}\n", handler.first, handler.second->help_text_));
     }
+    code = Http::Code::NotFound;
   }
 
   return code;
 }
 
+Http::Code AdminImpl::handlerAdminHome(const std::string&, Http::HeaderMap& header_map,
+                                       Buffer::Instance& response) {
+  header_map.addReferenceKey(Http::Headers::get().ContentType, "text/html; charset=UTF-8");
+
+  // Prefix order is used during searching, but for printing do them in alpha order.
+  std::map<std::string, const UrlHandler*> sorted_handlers;
+  for (const UrlHandler& handler : handlers_) {
+    sorted_handlers[handler.prefix_] = &handler;
+  }
+
+  response.add(R"(
+<head>
+  <link rel='shortcut icon' type='image/png' href='https://www.envoyproxy.io/img/favicon.ico'/>
+  <style>
+    .home-table {
+      font-family: sans-serif;
+      font-size: medium;
+      border-collapse: collapse;
+    }
+
+    .home-row:nth-child(even) {
+      background-color: #dddddd;
+    }
+
+    .home-data {
+      border: 1px solid #dddddd;
+      text-align: left;
+      padding: 8px;
+    }
+  </style>
+</head>
+<body>
+  <table class='home-table'>
+    <thead>
+      <th class='home-data'>Command</th>
+      <th class='home-data'>Description</th>
+     </thead>
+     <tbody>
+)");
+
+  for (auto handler : sorted_handlers) {
+    // Handlers are all specified by statically above, and are thus trusted and do
+    // not require escaping.
+    response.add(fmt::format("<tr class='home-row'><td class='home-data'><a href='{}'>{}</a></td>"
+                             "<td class='home-data'>{}</td></tr>\n",
+                             handler.first, handler.first,
+                             Html::Utility::sanitize(handler.second->help_text_)));
+  }
+  response.add(R"(
+    </tbody>
+  </table>
+</body>
+)");
+  return Http::Code::OK;
+}
+
 const Network::Address::Instance& AdminImpl::localAddress() {
   return *server_.localInfo().address();
 }
 
 bool AdminImpl::addHandler(const std::string& prefix, const std::string& help_text,
                            HandlerCb callback, bool removable) {
+  // Sanitize prefix and help_text to ensure no XSS can be injected, as
+  // we are injecting these strings into HTML that runs in a domain that
+  // can mutate Envoy server state. Also rule out some characters that
+  // make no sense as part of a URL path: ? and :.
+  //
+  // TODO(jmarantz): replace with absl definitive absl implementation
```

Comment:
Do we have this yet in public absl?

Reply:
actually I'll just remove the TODO.  The native STL seems fine.